### PR TITLE
s3api: Implement CORS operations

### PR DESCRIPTION
**Pull request by fvennetier**
Implement DeleteBucketCors, GetBucketCors and PutBucketCors S3 operations, along with OPTIONS requests on buckets and objects.

### CODE REVIEW

File: swift/common/middleware/s3api/controllers/cors.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+
+def match(pattern, value):
+    '''helper function for wildcard'''
+    if '*' not in pattern:
+        return pattern == value
+    # protect dot as we keep them as is
+    pattern = pattern.replace('.', '\\.')
+    pattern = '^' + pattern.replace('*', '.*') + '$'
+    return re.match(pattern, value) is not None
```
Comment (AymericDu) [inline swift/common/middleware/s3api/controllers/cors.py]: We should protect all non-alphanumerics, not just the dot.
```suggestion
    pattern_parts = pattern.split('*')
    if len(pattern_parts) == 1:
        return pattern == value
    # protect all non-alphanumerics (except wildcard) as we keep them as is
    regex_parts = list()
    for pattern_part in pattern_parts:
        regex_parts.append(re.escape(pattern_part))
    regex = '^' + '.*'.join(regex_parts) + '$'
    return re.match(regex, value) is not None
```

File: swift/common/middleware/s3api/controllers/cors.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    if not body:
+        return None
+    data = fromstring(body, "CorsConfiguration")
+
+    # we have to iterate over each to find matching origin
+    # whe have to manage wildcard in domain
+    rules = data.findall('CORSRule')
+    for rule in rules:
+        item = rule.find('AllowedOrigin')
+        if match(item.text, origin) or item.text == '*':
```
Comment (AymericDu) [inline swift/common/middleware/s3api/controllers/cors.py]: `item.text == '*'` is already checked with the `match` function.
```suggestion
        if match(item.text, origin):
```

File: swift/common/middleware/s3api/controllers/cors.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+                        allowed = [x.text.lower()
+                                   for x in rule.findall('AllowedHeader')]
+
+                        # manage * as well for headers
+                        hdrs = [x.lower().strip() for x in hdrs.split(',')]
+                        if '*' not in allowed \
+                                and not all([hdr in allowed for hdr in hdrs]):
+                            # some requested headers are not found
+                            continue
+                    return rule
```
Comment (AymericDu) [inline swift/common/middleware/s3api/controllers/cors.py]: Each `AllowedHeader` string in the rule can contain at most one * wildcard character. (https://docs.aws.amazon.com/AmazonS3/latest/dev/cors.html#cors-allowed-headers)
Also, if headers are not allowed, there is no point in continuing to check with other methods, the result will be the same.
```suggestion
            # check AllowedMethod
            methods = rule.findall('AllowedMethod')
            for method in methods:
                if method.text != method:
                    continue

                headers = req.headers.get('Access-Control-Request-Headers')
                if not headers:
                    return rule

                # check AllowedHeader
                headers = [x.lower().strip() for x in headers.split(',')]
                allowed_headers = [x.text.lower()
                                   for x in rule.findall('AllowedHeader')]
                for header in headers:
                    header_match = False
                    for allowed_header in allowed_headers:
                        if match(allowed_header, header):
                            header_match = True
                            break
                    if not header_match:
                        break
                else:
                    # all requested headers are allowed
                    return rule
                # some requested headers are not allowed
                break
```

File: swift/common/middleware/s3api/controllers/cors.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+def cors_fill_headers(req, resp, rule):
+    def set_header_if_item(hdr, tag):
+        val = rule.find(tag)
+        if val is not None:
+            resp.headers[hdr] = val.text
+
+    def set_header_if_items(hdr, tag):
+        vals = [m.text for m in rule.findall(tag)]
+        if vals:
+            resp.headers[hdr] = ', '.join(vals)
```
Comment (AymericDu) [inline swift/common/middleware/s3api/controllers/cors.py]: I think the space isn't useful.

Comment (fvennetier) [reply swift/common/middleware/s3api/controllers/cors.py]: Fixed.

File: swift/common/middleware/s3api/controllers/cors.py (hunk 1 above)
Comment (fvennetier) [reply swift/common/middleware/s3api/controllers/cors.py]: Thanks for the suggestion. I wrote a shorter version.

File: swift/common/middleware/s3api/controllers/cors.py (hunk 2 above)
Comment (fvennetier) [reply swift/common/middleware/s3api/controllers/cors.py]: Fixed.

**Review by AymericDu — APPROVED** [review]

**Review by murlock — APPROVED** [review]

**Pull request merged**